### PR TITLE
Add ETB trace library for nRF91

### PR DESCRIPTION
**Pull request by jtguggedal**
* Add library to capture instruction traces and store them to the Embedded Trace Buffer, ETB.
* Add support for storing captured ETB traces in coredump captured by Memfault SDK.
* Add support for ETB traces to the Memfault sample.

### CODE REVIEW

**Review by divipillai — APPROVED** [review]

File: subsys/debug/etb_trace/etb_trace.c [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+	SET_REG(ETM_TRCPRGCTLR, 0);
+
+	/* Wait until ETM is idle and PM is stable */
+	while ((GET_REG(ETM_TRCSTATR) & (BIT(1) | BIT(0))) != (BIT(1) | BIT(0))) {
+	}
+
+	/* Configure the ETM */
+	SET_REG(ETM_TRCCONFIGR, etm_trcconfigr);
+
+	/* The trace unit can not stall the processor for instruction traces, at the risk of
```
Comment (balaji-nordic) [inline subsys/debug/etb_trace/etb_trace.c:34]:
```suggestion
	/* Disable stalling of CPU as the trace unit can not stall the processor for instruction traces, at the risk of
```

File: subsys/debug/etb_trace/etb_trace.c [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	SET_REG(ETM_TRCTRACEIDR, etm_trctraceidr);
+
+	/* Bit 0: Enable event 0.
+	 * Bit 9: Indicates the current status of the start/stop logic.
+	 * Bit 10: Always trace reset exceptions.
+	 * Bit 11: Always trace system error exceptions.
+	 */
+	SET_REG(ETM_TRCVICTLR, BIT(11) | BIT(10) | BIT(9) | BIT(0));
+
+	/* No events are configured */
```
Comment (balaji-nordic) [inline subsys/debug/etb_trace/etb_trace.c:52]: Some more explanation would be nice. What are these events? and why should they not be configured?

File: subsys/debug/etb_trace/etb_trace.c [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	while ((GET_REG(ETB_FFSR) & BIT(1)) == 0) {
+	}
+
+	/* Enable formatter in continuous mode */
+	SET_REG(ETB_FFCR, BIT(1) | BIT(0));
+
+	/* Enable ETB */
+	SET_REG(ETB_CTL, 0x1);
+
+	while ((GET_REG(ETB_FFSR) & BIT(1)) != 0) {
```
Comment (balaji-nordic) [inline subsys/debug/etb_trace/etb_trace.c:118]: Whats happening here?

File: subsys/debug/etb_trace/etb_trace_pm.c [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * @brief Turn off debug unit and tracing if going to sleep for a long period of time.
+ *
+ * @return always true to allow going to idle
+ */
+bool z_arm_on_enter_cpu_idle(void)
+{
+	/* Stop only if traces are enabled */
+	if ((GET_REG(ETB_FFSR) & BIT(1)) != 0) {
```
Comment (balaji-nordic) [inline subsys/debug/etb_trace/etb_trace_pm.c:37]: May be create an API in the etb_trace.c for this?

File: subsys/debug/etb_trace/etb_trace_pm.c [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+		return true;
+	}
+
+	if (_kernel.idle > k_ms_to_ticks_ceil32(CONFIG_ETB_TRACE_PM_MIN_IDLE_TIME_MS)) {
+		etb_trace_stop();
+		NRF_TAD_S->TASKS_CLOCKSTOP = TAD_TASKS_CLOCKSTOP_TASKS_CLOCKSTOP_Msk;
```
Comment (balaji-nordic) [inline subsys/debug/etb_trace/etb_trace_pm.c:43]: Dont we need to start the clock in `sys_clock_idle_exit()`?

File: subsys/debug/etb_trace/etb_trace_private.h [hunk 6]
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
```
Comment (balaji-nordic) [inline subsys/debug/etb_trace/etb_trace_private.h:6]: Explain why this file exists and perhaps the reasoning behind the prefixes (`CS`, `ATB_1` and `ATB_2`)

**Review by balaji-nordic — CHANGES_REQUESTED** [review]

File: doc/nrf/libraries/debug/etb_trace.rst [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+Configuration
+*************
+
+To enable the library, use the :kconfig:option:`CONFIG_ETB_TRACE` Kconfig option.
+By default, the library automatically starts tracing at the earliest stage in the Zephyr boot sequence to capture instructions as early as possible.
+You can choose to start tracing in the application by disabling the :kconfig:option:`CONFIG_ETB_TRACE_SYS_INIT` Kconfig option.
+
+To capture ETM-generated traces and store them in ETB, the SoC's trace and debug system must be enabled.
+This increases power consumption significantly.
```
[the remark this answers is not in the document]
Comment (maxd-nordic) [reply doc/nrf/libraries/debug/etb_trace.rst:28]: Note: It was 4-6mA in practice.

File: include/debug/etb_trace.h [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ */
+void etb_trace_stop(void);
+
+/**
+ * @brief Retrieve ETB trace data to buffer
+ *
+ * @param[out] buf output buffer
+ * @param[in] buf_size size of output buffer
+ * @return size_t number of bytes written
```
[the remark this answers is not in the document]
Comment (maxd-nordic) [reply include/debug/etb_trace.h:43]: I think buf doesn't need to be volatile

File: subsys/debug/etb_trace/etb_trace.c [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/init.h>
+#include <debug/etb_trace.h>
+
+#include "etb_trace_private.h"
+
+/* Declaring as static variables to make them appear with the correct value in ELF file
```
[the remark this answers is not in the document]
Comment (maxd-nordic) [reply subsys/debug/etb_trace/etb_trace.c:13]: This is meant to say: We need to be able to read them from the ELF file so we can feed the decoder with the correct config.

File: subsys/debug/etb_trace/etb_trace.c [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+ * for simplified trace processing.
+ */
+
+/* Bit 3: Branch broadcast mode enabled. */
+static volatile uint32_t etm_trcconfigr = BIT(3);
+/* Set trace ID to 0x10  */
+static volatile uint32_t etm_trctraceidr = 0x10;
```
[the remark this answers is not in the document]
Comment (maxd-nordic) [reply subsys/debug/etb_trace/etb_trace.c:20]: I assume the Trace ID is somewhat arbitrary, but needs to match the ATB config below. This is a bit difficult to document before we have official documentation for this feature.

File: subsys/debug/etb_trace/etb_trace.c [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+
+static void etm_init(void)
+{
+	/* Disable ETM to allow configuration */
+	SET_REG(ETM_TRCPRGCTLR, 0);
+
+	/* Wait until ETM is idle and PM is stable */
```
[the remark this answers is not in the document]
Comment (maxd-nordic) [reply subsys/debug/etb_trace/etb_trace.c:27]: this waits until the PMSTABLE, bit [1] and IDLE, bit [0] are set. PM is programmer's model

File: subsys/debug/etb_trace/etb_trace.c [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+	SET_REG(ETM_TRCPRGCTLR, 0);
+}
+
+static void atb_init(void)
+{
+	/* ATB replicator */
+	CS_UNLOCK(ATB_REPLICATOR_BASE_ADDR);
+
+	/* ID filter for master port 0 */
+	SET_REG(ATB_REPLICATOR_IDFILTER0, 0xFFFFFFFFUL);
```
[the remark this answers is not in the document]
Comment (maxd-nordic) [reply subsys/debug/etb_trace/etb_trace.c:77]: ATB related stuff is not publicly documented yet. Cannot fix.

File: subsys/debug/etb_trace/etb_trace.c [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+	/* ID filter for master port 1, allowing ETM traces from CM33 to ETB */
+	SET_REG(ATB_REPLICATOR_IDFILTER1, 0xFFFFFFFDUL);
```
[the remark this answers is not in the document]
Comment (maxd-nordic) [reply subsys/debug/etb_trace/etb_trace.c:79]: ATB related stuff is not publicly documented yet. Cannot fix.

File: subsys/debug/etb_trace/etb_trace.c [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+
+	CS_LOCK(ATB_REPLICATOR_BASE_ADDR);
+
+	/* ATB funnel 1 */
+	CS_UNLOCK(ATB_1_BASE_ADDR);
+
+	/* Set pririty 1 for ports 0 and 1 */
+	SET_REG(ATB_1_PRIO, 0x00000009UL);
```
[the remark this answers is not in the document]
Comment (maxd-nordic) [reply subsys/debug/etb_trace/etb_trace.c:87]: ATB related stuff is not publicly documented yet. Cannot fix.

File: subsys/debug/etb_trace/etb_trace.c (hunk 2 above)
Comment (maxd-nordic) [reply subsys/debug/etb_trace/etb_trace.c:57]: According to the Technical Reference Manual, we have to write these registers as part of initialization. They configure features we don't use, so we just write 0 to disable them. Is it really necessary to go into details about these unused features?

File: subsys/debug/etb_trace/etb_trace_pm.c (hunk 5 above)
Comment (maxd-nordic) [reply subsys/debug/etb_trace/etb_trace_pm.c:43]: we do that along with a full initialization with etb_trace_start()

File: subsys/debug/etb_trace/etb_trace_pm.c (hunk 4 above)
Comment (maxd-nordic) [reply subsys/debug/etb_trace/etb_trace_pm.c:37]: done

File: subsys/debug/etb_trace/etb_trace.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (gregersrygg) [reply subsys/debug/etb_trace/etb_trace.c:77]: Why can we explain what it's doing in the comment above then? `REG_ID_FILTER_M0` would work for me.

**Review by gregersrygg — APPROVED** [review]
> ATB related stuff is not publicly documented yet. Cannot fix.

I would still like a define/const for the register addresses. If it can be in a comment just above it can also be in a symbol.

File: subsys/debug/etb_trace/etb_trace_private.h [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+#define ETM_TRCSTALLCTLR		(ETM_BASE_ADDR + 0x02C) /* Stall Control Register */
+#define ETM_TRCTSCTLR			(ETM_BASE_ADDR + 0x030) /* Timestamp Control Register */
+#define ETM_TRCTRACEIDR			(ETM_BASE_ADDR + 0x040) /* Trace ID Register */
+#define ETM_TRCVICTLR			(ETM_BASE_ADDR + 0x080) /* ViewInst Main Control Register */
+#define ETM_TRCEVENTCTL0R		(ETM_BASE_ADDR + 0x020) /* Event Control 0 Register */
+#define ETM_TRCEVENTCTL1R		(ETM_BASE_ADDR + 0x024) /* Event Control 1 Register */
+#define ETM_TRCPDSR			(ETM_BASE_ADDR + 0x314) /* Power down status register */
+
+#define ETM_TRCSTATR_IDLE		BIT(0)
+#define ETM_TRCSTATR_PMSTABLE		BIT(1)
```
Comment (balaji-nordic) [inline subsys/debug/etb_trace/etb_trace_private.h:68]:
```suggestion
#define ETM_TRCSTATR_READY		BIT(1)
```

File: subsys/debug/etb_trace/etb_trace.c [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+	SET_REG(ETB_CTL, 0);
+
+	/* Wait for ETB formatter to stop */
+	while (!ETB_FORMATTER_STOPPED) {
+	}
+
+	CS_LOCK(ETB_BASE_ADDR);
+}
+
+static void debug_init(void)
```
Comment (balaji-nordic) [inline subsys/debug/etb_trace/etb_trace.c:144]:
```suggestion
static void trace_debug_clock_start(void)
```

File: subsys/debug/etb_trace/etb_trace_pm.c (hunk 5 above)
Comment (balaji-nordic) [reply subsys/debug/etb_trace/etb_trace_pm.c]: Then we should move this line to `etb_trace_stop()`. Perhaps have a static funciton in etb_trace.c (just like a counter part to the current `debug_init()`)

File: subsys/debug/etb_trace/etb_trace.c [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+{
+	debug_init();
+	atb_init();
+	etb_init();
+	etm_init();
+}
+
+void etb_trace_stop(void)
+{
+	etm_stop();
```
Comment (balaji-nordic) [inline subsys/debug/etb_trace/etb_trace.c:159]: No need to stop atb?

File: subsys/debug/etb_trace/etb_trace_lp.c [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+/**
+ * @brief Turn off debug unit and tracing if going to sleep for a long period of time.
+ *
+ * @return always true to allow going to idle
+ */
+bool z_arm_on_enter_cpu_idle(void)
+{
+	/* Stop only if traces are enabled */
```
Comment (balaji-nordic) [inline subsys/debug/etb_trace/etb_trace_lp.c:36]: The comment and the code do not align well now. Perhaps the `ETB_FORMATTER_STOPPED` should be called `ETB_TRACE_DISABLED`?

Or may be check the state of `trace_stopped`?

**Review by balaji-nordic — CHANGES_REQUESTED** [review]

File: subsys/debug/etb_trace/etb_trace_private.h [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+#define ETM_TRCCCCTLR			(ETM_BASE_ADDR + 0x038) /* Cycle Count Control Register */
+#define ETM_TRCSTALLCTLR		(ETM_BASE_ADDR + 0x02C) /* Stall Control Register */
+#define ETM_TRCTSCTLR			(ETM_BASE_ADDR + 0x030) /* Timestamp Control Register */
+#define ETM_TRCTRACEIDR			(ETM_BASE_ADDR + 0x040) /* Trace ID Register */
+#define ETM_TRCVICTLR			(ETM_BASE_ADDR + 0x080) /* ViewInst Main Control Register */
+#define ETM_TRCEVENTCTL0R		(ETM_BASE_ADDR + 0x020) /* Event Control 0 Register */
+#define ETM_TRCEVENTCTL1R		(ETM_BASE_ADDR + 0x024) /* Event Control 1 Register */
+#define ETM_TRCPDSR			(ETM_BASE_ADDR + 0x314) /* Power down status register */
+
+#define ETM_TRCSTATR_IDLE		BIT(0)
```
[the remark this answers is not in the document]
Comment (maxd-nordic) [reply subsys/debug/etb_trace/etb_trace_private.h:68]: ETMSR != TRCSTATR https://developer.arm.com/documentation/ddi0601/2020-12/External-Registers/TRCSTATR--Trace-Status-Register?lang=en

File: subsys/debug/etb_trace/etb_trace.c [hunk 20]
```diff
@@ -0,0 +1,194 @@
...
+/* Set trace ID to 0x10  */
+static volatile uint32_t etm_trctraceidr = 0x10;
+
+static void etm_init(void)
+{
+	/* Disable ETM to allow configuration */
+	SET_REG(ETM_TRCPRGCTLR, 0);
+
+	/* Wait until ETM is idle and programmer's model is stable */
+	while ((GET_REG(ETM_TRCSTATR) & (ETM_TRCSTATR_PMSTABLE | ETM_TRCSTATR_IDLE))
```
[the remark this answers is not in the document]
Comment (maxd-nordic) [reply subsys/debug/etb_trace/etb_trace.c:28]: This is a different register.

File: subsys/debug/etb_trace/etb_trace.c (hunk 17 above)
Comment (maxd-nordic) [reply subsys/debug/etb_trace/etb_trace.c:159]: not for power saving - I think ATB just sleeps if there are no messages from the ETM or ITM.

File: subsys/debug/etb_trace/etb_trace.c [hunk 21]
```diff
@@ -0,0 +1,196 @@
...
+}
+
+size_t etb_data_get(uint32_t *buf, size_t buf_size)
+{
+	size_t i = 0;
+
+	if (buf_size == 0) {
+		return -EINVAL;
+	}
+
```
[the remark this answers is not in the document]
Comment (maxd-nordic) [reply subsys/debug/etb_trace/etb_trace.c:170]: done

File: include/debug/etb_trace.h [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ */
+void etb_trace_stop(void);
+
+/**
+ * @brief Retrieve ETB trace data to buffer
+ *
+ * @param[out] buf Output buffer
+ * @param[in] buf_size Size of output buffer in bytes
+ * @return size_t Number of bytes of trace data returned in the buf, a negative
```
[the remark this answers is not in the document]
Comment (maxd-nordic) [reply include/debug/etb_trace.h:43]: done

**Comment by maxd-nordic** [comment]
> Then we should move this line to etb_trace_stop(). Perhaps have a static funciton in etb_trace.c (just like a counter part to the current debug_init())

I wouldn't do that since it's part of the experimental low power stuff. There is no guarantee this is safe to do.

File: subsys/debug/etb_trace/etb_trace.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (maxd-nordic) [reply subsys/debug/etb_trace/etb_trace.c:20]: If I change them here, I'd have to change the scripts as well. Doesn't seem to be worth it.

File: doc/nrf/libraries/debug/etb_trace.rst [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+ETB trace
+#########
+
+.. contents::
+   :local:
+   :depth: 2
+
+The experimental Embedded Trace Buffer (ETB) library enables the tracing of executed instructions and storing them in a dedicated RAM buffer.
+Traces are generated by the core's Embedded Trace Macrocell (ETM) unit and are captured without CPU intervention.
+Currently, the nRF91 series SoCs are supported.
```
[the remark this answers is not in the document]
Comment (simtind) [reply doc/nrf/libraries/debug/etb_trace.rst]: NRF91 will soon also be SOCs
NRF91 SIPs are all based on the nrf9120 soc and it's not likely we'll get any other SOCs in this series. If we do it's not likely that etb will be dropped as tracing is fairly fundamental / cheap in such a big device.

[the remark this answers is not in the document]
Comment (simtind) [reply doc/nrf/libraries/debug/etb_trace.rst]: Changing to 9160 also ensures you will very quickly outdate this doc once the other 91 series devices are launched.

File: doc/nrf/libraries/debug/etb_trace.rst [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+To capture ETM-generated traces and store them in ETB, the SoC's trace and debug system must be enabled.
+This increases power consumption by serval milliamps.
+The library provides the option (:kconfig:option:`CONFIG_ETB_TRACE_LOW_POWER`) to reduce the power consumption by disabling the debug system when the CPU goes to idle.
+This is a feature that requires code to be executed both on entry and exit of the CPU idle state and will introduce latency to the system.
+It might also have other unknown side effects on the system.
+
+Usage
+*****
+
+To initialize and start tracing, call the :c:func:`etb_trace_start` function.
```
Comment (peknis) [inline doc/nrf/libraries/debug/etb_trace.rst:36]:
```suggestion
To initialize the library and start tracing, call the :c:func:`etb_trace_start` function.
```

File: doc/nrf/libraries/debug/etb_trace.rst [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+The function performs the following:
+
+#. Configures the ETM.
+#. Sets up the ATB funnels and replicator.
```
Comment (peknis) [inline doc/nrf/libraries/debug/etb_trace.rst:40]: Have we introduced ATB somewhere? If not, write it open.

File: doc/nrf/libraries/debug/etb_trace.rst [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+#. Enables the ETB.
+
+Traces are now written continuously to the ETB RAM.
+
+To stop trace capture, use the :c:func:`etb_trace_stop` function.
+
+The captured traces can now be read out using the :c:func:`etb_data_get` function.
+The maximum size of data the ETB buffer can hold is 2 KB.
```
Comment (peknis) [inline doc/nrf/libraries/debug/etb_trace.rst:48]:
```suggestion
The ETB buffer can hold the maximum of 2 KB of data.
```

Comment (peknis) [inline doc/nrf/libraries/debug/etb_trace.rst:48]:
```suggestion
The ETB buffer can hold the maximum of 2 KB of data.
```

File: include/debug/etb_trace.h [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+ */
+void etb_trace_stop(void);
+
+/**
+ * @brief Retrieve ETB trace data to buffer
+ *
+ * @param[out] buf Output buffer
+ * @param[in] buf_size Size of output buffer in bytes
+ * @retval Number of bytes of trace data returned in the buf
+ * @retval -EINVAL On invalid input parameters
```
Comment (peknis) [inline include/debug/etb_trace.h:44]:
```suggestion
 * @brief Retrieve ETB trace data to buffer.
 *
 * @param[out] buf Output buffer.
 * @param[in] buf_size Size of output buffer in bytes.
 * @retval Number of bytes of trace data returned in buf.
 * @retval -EINVAL On invalid input parameters.
```

Comment (peknis) [inline include/debug/etb_trace.h:44]:
```suggestion
 * @brief Retrieve ETB trace data to buffer.
 *
 * @param[out] buf Output buffer.
 * @param[in] buf_size Size of output buffer in bytes.
 * @retval Number of bytes of trace data returned in buf.
 * @retval -EINVAL On invalid input parameters.
```

**Review by peknis — CHANGES_REQUESTED** [review]

File: doc/nrf/libraries/debug/etb_trace.rst (hunk 23 above)
[the remark this answers is not in the document]
Comment (jhn-nordic) [reply doc/nrf/libraries/debug/etb_trace.rst]: Agree with @simtind

File: subsys/debug/etb_trace/etb_trace.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (jhn-nordic) [reply subsys/debug/etb_trace/etb_trace.c:77]: We can keep these things as is until it has a proper name in docs.

[the remark this answers is not in the document]
Comment (maxd-nordic) [reply subsys/debug/etb_trace/etb_trace.c:77]: My idea about this: If we come up with names for those things here, there will be a mismatch when the docs are released. That's much worse than keeping the raw values.

File: subsys/debug/etb_trace/etb_trace_pm.c (hunk 5 above)
Comment (jhn-nordic) [reply subsys/debug/etb_trace/etb_trace_pm.c]: Agree with @maxd-nordic here. Lets keep this as is.

File: doc/nrf/libraries/debug/etb_trace.rst [hunk 28]
```diff
@@ -0,0 +1,59 @@
...
+ETB trace
+#########
+
+.. contents::
+   :local:
+   :depth: 2
+
+The experimental Embedded Trace Buffer (ETB) library enables the tracing of executed instructions and storing them in a dedicated RAM buffer.
+Traces are generated by the core's Embedded Trace Macrocell (ETM) unit and are captured without CPU intervention.
+Currently, the nRF9160 is supported.
```
[the remark this answers is not in the document]
Comment (maxd-nordic) [reply doc/nrf/libraries/debug/etb_trace.rst:12]: I'll change it to nRF91 series.

File: doc/nrf/libraries/debug/etb_trace.rst (hunk 23 above)
[the remark this answers is not in the document]
Comment (maxd-nordic) [reply doc/nrf/libraries/debug/etb_trace.rst]: I'll change it to nRF91 series.

File: doc/nrf/libraries/debug/etb_trace.rst (hunk 28 above)
[the remark this answers is not in the document]
Comment (jhn-nordic) [reply doc/nrf/libraries/debug/etb_trace.rst:12]: Can we say the nRF91 series?

File: doc/nrf/libraries/debug/etb_trace.rst (hunk 23 above)
[the remark this answers is not in the document]
Comment (peknis) [reply doc/nrf/libraries/debug/etb_trace.rst]: nRF91 Series devices.

File: doc/nrf/libraries/debug/etb_trace.rst (hunk 25 above)
Comment (peknis) [reply doc/nrf/libraries/debug/etb_trace.rst:40]: The glossary in AMBA™ 3 ATB Protocol spec says it is **Advanced Trace Bus**

**Review by peknis — APPROVED** [review]

File: subsys/debug/etb_trace/etb_trace_private.h (hunk 19 above)
[the remark this answers is not in the document]
Comment (balaji-nordic) [reply subsys/debug/etb_trace/etb_trace_private.h:68]: ahah. thanks for pointing that out

File: subsys/debug/etb_trace/etb_trace.c (hunk 20 above)
[the remark this answers is not in the document]
Comment (balaji-nordic) [reply subsys/debug/etb_trace/etb_trace.c:28]: yes the names are confusing me. Can we just remove the ETM prefix then?

File: subsys/debug/etb_trace/etb_trace_pm.c (hunk 5 above)
Comment (balaji-nordic) [reply subsys/debug/etb_trace/etb_trace_pm.c]: oh then should we remove the code that stops the clock here?

**Review by balaji-nordic — APPROVED** [review]

**Review by simtind — APPROVED** [review]

**Pull request merged**